### PR TITLE
Add orchestration property tests (#72)

### LINKED ISSUE #72: Add Hypothesis property-based tests for episodic/orchestration/
## Background

PR #69 introduced the `episodic/orchestration/` package, including enum normalisation, model-tier validation, JSON parsing error preservation, and a LangGraph state-machine. The review identified that the current test suite uses only deterministic unit tests with hand-picked payloads, and flagged this as a gap for exhaustive edge-case exploration.

`hypothesis` is an accepted project dependency.

## Work required

Add property-based tests using Hypothesis to:

1. **Enum normalisation** — verify `GenerationOrchestrationConfig` correctly normalises arbitrary mixes of `ActionKind` enum values and their string equivalents across the full input space.
2. **Model-tier validation rejection** — verify `ShowNotesToolExecutor.execute()` rejects every `ModelTier` value other than `ModelTier.EXECUTION`.
3. **JSON parsing error preservation** — verify `StructuredGenerationPlanner` raises `PlanningResponseFormatError` for all malformed/invalid plan payloads, not just hand-picked examples.
4. **LangGraph state-machine ordering invariants** — verify that across arbitrary valid `GenerationOrchestrationRequest` inputs the `plan → execute → finish` node ordering is always respected and that aggregated token counts are always non-negative.

## References

- PR: https://github.com/leynos/episodic/pull/69
- Review comment: https://github.com/leynos/episodic/pull/69#issuecomment-3168199131
- Implementation: `episodic/orchestration/generation.py`, `episodic/orchestration/langgraph.py`
- Existing tests: `tests/test_generation_orchestration.py`, `tests/test_generation_orchestration_langgraph.py`

Raised by @leynos.

### PR DESCRIPTION
**Pull request by leynos**
## Summary

This branch adds Hypothesis coverage for the structured generation orchestration contracts raised in issue #72. It exercises enum normalisation across mixed `ActionKind` inputs, rejects every non-execution `ModelTier`, broadens planner format-error preservation across malformed JSON and invalid plan objects, and checks LangGraph ordering and token aggregation invariants.

Review follow-up changes expand the module documentation, make graph event recording an explicit injected dependency, bound generated malformed payloads, and pin representative `PlanningResponseFormatError` messages with Syrupy snapshots.

Closes #72.

## Review walkthrough

- Start with [tests/test_orchestration_properties.py](https://github.com/leynos/episodic/blob/issue-72-hypothesis-property-based-tests-for-episodic-orchestration/tests/test_orchestration_properties.py) for the Hypothesis strategies and orchestration invariants.
- Review [tests/test_generation_orchestration_snapshots.py](https://github.com/leynos/episodic/blob/issue-72-hypothesis-property-based-tests-for-episodic-orchestration/tests/test_generation_orchestration_snapshots.py) for representative planner error-message snapshots.
- Finish with [tests/__snapshots__/test_generation_orchestration_snapshots.ambr](https://github.com/leynos/episodic/blob/issue-72-hypothesis-property-based-tests-for-episodic-orchestration/tests/__snapshots__/test_generation_orchestration_snapshots.ambr) for the pinned diagnostics.

## Validation

- `make check-fmt`: passed
- `make lint`: passed
- `make test PYTEST_XDIST_WORKERS=1`: passed, 481 passed, 3 skipped
- `make typecheck`: passed

## Notes

`hypothesis` was already present in the dev dependency group, so no dependency or lockfile change was required. One full-suite run hit a transient `py-pglite` fixture setup timeout in an unrelated reference-document pagination case; the isolated case passed immediately, and the subsequent full-suite retry passed.

### CODE REVIEW

**Comment by leynos** [comment]
@coderabbitai Have the following now been resolved?

#### tests/test_orchestration_properties.py (1)

`422-424`: _⚠️ Potential issue_ | _🟠 Major_ | _⚡ Quick win_
**Replace probabilistic tier sampling with deterministic exhaustive coverage.**
Hypothesis `st.sampled_from()` with `@settings(max_examples=len(ModelTier))` does not guarantee every non-execution tier is tested—values are drawn with replacement, and duplicates can skip enum members entirely. The test name claims to reject "all non-execution tiers", yet the implementation provides probabilistic rather than exhaustive coverage. Use `pytest.mark.parametrize` to guarantee all tiers are exercised.

##### Proposed fix

```diff
-@given(model_tier=st.sampled_from([t for t in ModelTier if t != ModelTier.EXECUTION]))
-@settings(max_examples=len(ModelTier))
-@pytest.mark.asyncio
-async def test_planned_action_model_tier_rejection_for_all_non_execution_tiers(
-    model_tier: ModelTier,
-) -> None:
+@pytest.mark.parametrize(
+    ("model_tier",),
+    [(tier,) for tier in ModelTier if tier is not ModelTier.EXECUTION],
+)
+@pytest.mark.asyncio
+async def test_planned_action_model_tier_rejection_for_all_non_execution_tiers(
+    model_tier: ModelTier,
+) -> None:
```

Coding guidelines (`**/test_*.py`) require "parametrize broadly" for proper test idioms.

##### 🤖 Detailed instructions

Verify each finding against current code. Fix only still-valid issues, skip the
rest with a brief reason, keep changes minimal, and validate.

In `@tests/test_orchestration_properties.py` around lines 422 - 424, Replace the
probabilistic Hypothesis-based sampling with deterministic parametric coverage:
remove the `@given`(...) and `@settings`(...) decorators and instead add a
pytest.mark.parametrize for the
test_planned_action_model_tier_rejection_for_all_non_execution_tiers test that
iterates every ModelTier member except ModelTier.EXECUTION (e.g., parametrize
("model_tier",) with [(tier,) for tier in ModelTier if tier is not
ModelTier.EXECUTION]); keep `@pytest.mark.asyncio` and the async test signature
unchanged so the test deterministically exercises all non-execution tiers.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 5 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Test test\_planned\_action\_model\_tier\_rejection uses `@given` with sampled\_from filtering to single value with max\_examples=10, generating identical test cases. Review requested parametrised coverage. | Replace `@given/`@settings with `@pytest.mark.parametrize` covering ModelTier excluding EXECUTION, per review comments and PR objectives. |
| Unit Architecture | ❌ Error | The `finish_callback` parameter accepts unvalidated code without explicit error handling. Fallibility is hidden from the API with no documented exception contract, allowing uncaught errors. | Document `finish_callback` fallibility contract explicitly. Surface exceptions or mark infallible in type/docs. Handle errors in `_run_finish_node`. Decouple event recording from test fakes; use test wrappers instead. |
| User-Facing Documentation | ⚠️ Warning | PR introduces a new optional `finish_callback` parameter to public API function `build_generation_orchestration_graph()`, but no documentation was updated or added. | Document the new `finish_callback` parameter in the function's docstring and in `docs/developers-guide.md` orchestration section, explaining its purpose, type signature, and usage context. |
| Developer Documentation | ⚠️ Warning | PR adds new public API parameter `finish_callback` and property tests for issue `#72`, but fails to document in `docs/developers-guide.md`. No design documents, ADRs, or roadmap updates. | Document `finish_callback` parameter and issue `#72` tests in orchestration section of `docs/developers-guide.md`. Update roadmap to mark issue `#72` complete. Consider adding an ADR if warranted. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Tier rejection test uses `@given/`@settings instead of parametrize as requested. finish\_callback untested with checkpoint\_port. | Use `@pytest.mark.parametrize` for exhaustive ModelTier coverage. Test finish\_callback in both direct-execute and suspend-before-execute graph paths. |
| Domain Architecture | ⚠️ Warning | finish\_callback parameter accepts GenerationGraphState, a LangGraph framework detail. Domain architecture requires callbacks to accept domain-shaped events, not framework structures. | Redefine finish\_callback to accept domain-level objects rather than LangGraph's GenerationGraphState implementation detail. |
| Observability | ⚠️ Warning | The finish\_callback invokes user code without exception handling or logging, creating an unobservable failure point in a critical path. | Wrap finish\_callback invocation in try/except with \_log\_event calls for success/failure. Document the callback contract and exception behaviour in the function docstring. |

**Comment by leynos** [comment]
@coderabbitai please create a new GitHub issue for the following:

## ⚠️ Warning 6 — Domain Architecture

**Status: Not resolved.** `finish_callback` exposes `GenerationGraphState` — a LangGraph framework type — in the public API of `build_generation_orchestration_graph`. This couples callers to the framework.

This is a design decision that likely warrants its own issue rather than being addressed inline. A GitHub issue does not currently exist for it.

Shall I open a GitHub issue to track replacing `finish_callback: Callable[[GenerationGraphState], None]` with a callback that accepts a domain-level result object (e.g. `GenerationOrchestrationResult`)? That would be the correct resolution per the Domain Architecture warning without blocking this PR.

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[episodic/orchestration/langgraph.py](https://github.com/leynos/episodic/pull/97/files/ba37458b3041940f1326025e6ba0fe05cfca37ab#diff-10b009ace75f20bef48d24feaa560ccc5d1e98c4ab864bacc4b1fc14db3c3de9)

Comment on file

~~~python
from episodic.orchestration._usage import build_generation_result

if typ.TYPE_CHECKING:
    import collections.abc as cabc
~~~

❌ New issue: [Lines of Code in a Single File](https://codescene.io/projects/76628/delta?repo-id=1455498&review-id=97&biomarker=Lines+of+Code+in+a+Single+File&filename=episodic%2Forchestration%2Flanggraph.py&method=)
This module has 610 lines of code, improve code health by reducing it to 400

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[episodic/orchestration/langgraph.py](https://github.com/leynos/episodic/pull/97/files/ba37458b3041940f1326025e6ba0fe05cfca37ab#diff-10b009ace75f20bef48d24feaa560ccc5d1e98c4ab864bacc4b1fc14db3c3de9)

Comment on file

~~~python
    planner: protocols.PlannerPort,
    tool_executor: protocols.ToolExecutorPort,
    checkpoint_port: protocols.CheckpointPort | None = None,
    finish_callback: cabc.Callable[[GenerationGraphState], None] | None = None,
~~~

❌ New issue: [Large Method](https://codescene.io/projects/76628/delta?repo-id=1455498&review-id=97&biomarker=Large+Method&filename=episodic%2Forchestration%2Flanggraph.py&method=build_generation_orchestration_graph)
build_generation_orchestration_graph has 96 lines, threshold = 70

File: docs/developers-guide.md [hunk 1]
```diff
@@ -702,6 +702,14 @@ Roadmap item `2.4.1` introduces a dedicated orchestration package in
 - `episodic/canonical/storage/workflow_checkpoints.py` contains the SQLAlchemy
   adapter for durable checkpoint persistence.
 
+`build_generation_orchestration_graph(...)` accepts an optional
+`finish_callback: Callable[[GenerationGraphState], None]` for observability and
+test event recording. The hook fires only on the direct
+`plan -> execute -> finish` path, after finish-node aggregation has produced an
+`orchestration_result` and before the graph returns. It is not invoked on the
+checkpoint suspend path. Callback exceptions are logged and propagated to the
+callback error log without replacing the already computed graph result.
```
[the remark this answers is not in the document]
Comment (leynos) [reply docs/developers-guide.md]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

File: tests/test_orchestration_planner_format_properties.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+        st.floats(allow_nan=False, allow_infinity=False),
+        st.lists(st.text(max_size=512), max_size=10),
+        st.text(max_size=512),
+    )
+)
+@settings(max_examples=50)
+@pytest.mark.asyncio
+async def test_planning_response_format_error_for_arbitrary_non_object_json(
+    noise: object,
+) -> None:
```
[the remark this answers is not in the document]
Comment (leynos) [reply tests/test_orchestration_planner_format_properties.py]: @coderabbitai Has this now been resolved in the latest commit?

Use codegraph analysis to determine your answer.

If this comment is now resolved, please mark it as such using the API. Otherwise, please provide an AI agent prompt for the remaining work to be done to address this comment.

**Comment by leynos** [comment]
@coderabbitai Have the following failed checks now been resolved?

If further work is required, please provide an AI agent prompt for the remaining work to be done to address these failures.

Do not treat warnings as optional or aspirational. Where a change is out of scope for this PR, propose a GitHub issue unless one exists already.

### ❌ Failed checks (2 errors, 5 warnings)

| Check name | Status | Explanation | Resolution |
| :-: | :-- | :-- | :-- |
| Testing (Overall) | ❌ Error | Test test\_planned\_action\_model\_tier\_rejection uses `@given` with sampled\_from filtering to single value with max\_examples=10, generating identical test cases. Review requested parametrised coverage. | Replace `@given/`@settings with `@pytest.mark.parametrize` covering ModelTier excluding EXECUTION, per review comments and PR objectives. |
| Unit Architecture | ❌ Error | The `finish_callback` parameter accepts unvalidated code without explicit error handling. Fallibility is hidden from the API with no documented exception contract, allowing uncaught errors. | Document `finish_callback` fallibility contract explicitly. Surface exceptions or mark infallible in type/docs. Handle errors in `_run_finish_node`. Decouple event recording from test fakes; use test wrappers instead. |
| User-Facing Documentation | ⚠️ Warning | PR introduces a new optional `finish_callback` parameter to public API function `build_generation_orchestration_graph()`, but no documentation was updated or added. | Document the new `finish_callback` parameter in the function's docstring and in `docs/developers-guide.md` orchestration section, explaining its purpose, type signature, and usage context. |
| Developer Documentation | ⚠️ Warning | PR adds new public API parameter `finish_callback` and property tests for issue `#72`, but fails to document in `docs/developers-guide.md`. No design documents, ADRs, or roadmap updates. | Document `finish_callback` parameter and issue `#72` tests in orchestration section of `docs/developers-guide.md`. Update roadmap to mark issue `#72` complete. Consider adding an ADR if warranted. |
| Testing (Unit And Behavioural) | ⚠️ Warning | Tier rejection test uses `@given/`@settings instead of parametrize as requested. finish\_callback untested with checkpoint\_port. | Use `@pytest.mark.parametrize` for exhaustive ModelTier coverage. Test finish\_callback in both direct-execute and suspend-before-execute graph paths. |
| Domain Architecture | ⚠️ Warning | finish\_callback parameter accepts GenerationGraphState, a LangGraph framework detail. Domain architecture requires callbacks to accept domain-shaped events, not framework structures. | Redefine finish\_callback to accept domain-level objects rather than LangGraph's GenerationGraphState implementation detail. |
| Observability | ⚠️ Warning | The finish\_callback invokes user code without exception handling or logging, creating an unobservable failure point in a critical path. | Wrap finish\_callback invocation in try/except with \_log\_event calls for success/failure. Document the callback contract and exception behaviour in the function docstring. |

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[episodic/orchestration/langgraph.py](https://github.com/leynos/episodic/pull/97/files/f7ba01f2c336fb05ef0b1e5db0a1097aea89c0c7#diff-10b009ace75f20bef48d24feaa560ccc5d1e98c4ab864bacc4b1fc14db3c3de9)

Comment on file

~~~python
from episodic.orchestration._usage import build_generation_result

if typ.TYPE_CHECKING:
    import collections.abc as cabc
~~~

❌ New issue: [Lines of Code in a Single File](https://codescene.io/projects/76628/delta?repo-id=1455498&review-id=97&biomarker=Lines+of+Code+in+a+Single+File&filename=episodic%2Forchestration%2Flanggraph.py&method=)
This module has 609 lines of code, improve code health by reducing it to 400

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[episodic/orchestration/langgraph.py](https://github.com/leynos/episodic/pull/97/files/f7ba01f2c336fb05ef0b1e5db0a1097aea89c0c7#diff-10b009ace75f20bef48d24feaa560ccc5d1e98c4ab864bacc4b1fc14db3c3de9)

Comment on lines +612 to +613

~~~python
    finish_callback: cabc.Callable[[dto.GenerationOrchestrationResult], None]
    | None = None,
~~~

❌ New issue: [Large Method](https://codescene.io/projects/76628/delta?repo-id=1455498&review-id=97&biomarker=Large+Method&filename=episodic%2Forchestration%2Flanggraph.py&method=build_generation_orchestration_graph)
build_generation_orchestration_graph has 95 lines, threshold = 70

**Comment by leynos** [comment]
@coderabbitai Please suggest a fix for this issue and supply a prompt for an AI coding agent to enable it to apply the fix. Include the file and symbol names indicated in the issue at the head of your response.

[tests/test_orchestration_graph_invariant_properties.py](https://github.com/leynos/episodic/pull/97/files/f7ba01f2c336fb05ef0b1e5db0a1097aea89c0c7#diff-8fe280115a5ffb909b98bb69277ce35889023ce53dbb1f597a57fd0d11f7b4b8)

Comment on lines +230 to +243

~~~python
async def test_finish_callback_is_invoked_in_direct_execute_path() -> None:
    """Direct execution invokes the finish callback with finished state."""
    observed_results: list[GenerationOrchestrationResult] = []
    graph = build_generation_orchestration_graph(
        planner=PropGraphPlanner(result=_planner_result()),
        tool_executor=PropGraphToolExecutor(result=_tool_result()),
        finish_callback=observed_results.append,
    )

    state = await graph.ainvoke(GenerationGraphState(request=_request()))

    assert len(observed_results) == 1
    assert observed_results[0] is not None
    assert state["orchestration_result"] == observed_results[0]
~~~

❌ New issue: [Code Duplication](https://codescene.io/projects/76628/delta?repo-id=1455498&review-id=97&biomarker=Code+Duplication&filename=tests%2Ftest_orchestration_graph_invariant_properties.py&method=)
The module contains 2 functions with similar structure: test_finish_callback_is_invoked_in_direct_execute_path,test_finish_callback_is_not_invoked_in_suspend_path
